common/archiver/provider/provider.go

```diff
@@ -0,0 +1,110 @@
+// THE SOFTWARE.
+
+package provider
+
+import (
+	"errors"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/archiver"
+	"github.com/uber/cadence/common/archiver/filestore"
+	"github.com/uber/cadence/common/log/tag"
+)
+
+var (
+	// ErrUnknownScheme is the error for unknown archiver scheme
+	ErrUnknownScheme = errors.New("unknown archiver scheme")
+	// ErrUnknownServiceName is the error for unknown cadence service name
+	ErrUnknownServiceName = errors.New("unknown service name")
+)
+
+type (
+	// Provider returns history or visiblity archiver based on the scheme and serviceName.
+	// The archiver for each combination of scheme and serviceName will be created only once and cached.
+	Provider interface {
+		GetHistoryArchiver(scheme string, serviceName string) (archiver.HistoryArchiver, error)
+		GetVisibilityArchiver(scheme string, serviceName string) (archiver.VisibilityArchiver, error)
+	}
+
+	// HistoryArchiverConfigs contain config for all implementations of the HistoryArchiver interface
+	HistoryArchiverConfigs struct {
+		FileStore *filestore.HistoryArchiverConfig
+	}
+
+	// VisibilityArchiverConfigs contain config for all implementations of the VisibilityArchiver interface
+	VisibilityArchiverConfigs struct {
+	}
+
+	provider struct {
+		container                 *archiver.BootstrapContainer
+		historyArchiverConfigs    *HistoryArchiverConfigs
+		visibilityArchiverConfigs *VisibilityArchiverConfigs
+
+		historyArchivers    map[string]archiver.HistoryArchiver
+		visibilityArchivers map[string]archiver.VisibilityArchiver
+	}
+)
+
+// NewArchiverProvider returns a new Archiver provider
+func NewArchiverProvider(
+	container *archiver.BootstrapContainer,
+	historyArchiverConfigs *HistoryArchiverConfigs,
+	visibilityArchiverConfigs *VisibilityArchiverConfigs,
+) Provider {
+	return &provider{
+		container:                 container,
+		historyArchiverConfigs:    historyArchiverConfigs,
+		visibilityArchiverConfigs: visibilityArchiverConfigs,
+	}
+}
+
+func (p *provider) GetHistoryArchiver(scheme, serviceName string) (archiver.HistoryArchiver, error) {
+	key := p.getArchiverKey(scheme, serviceName)
+	if archiver, ok := p.historyArchivers[key]; ok {
+		return archiver, nil
+	}
+
+	var componentTag tag.Tag
+	switch serviceName {
+	case common.HistoryServiceName:
+		componentTag = tag.ComponentTimerQueue
+	case common.WorkerServiceName:
+		componentTag = tag.ComponentArchiver
+	default:
+		return nil, ErrUnknownServiceName
+	}
+
+	switch scheme {
+	case filestore.URIScheme:
+		container := *p.container
+		container.Logger = container.Logger.WithTags(tag.Service(serviceName), componentTag)
+		return filestore.NewHistoryArchiver(container, p.historyArchiverConfigs.FileStore), nil
+	}
+	return nil, ErrUnknownScheme
+}
+
+func (p *provider) GetVisibilityArchiver(scheme, serviceName string) (archiver.VisibilityArchiver, error) {
+	return nil, errors.New("GetVisibilityArchiver is not implemented")
```

Comment:
Can we implement this. Everything can be implemented except the lowest level of filestore implementation that can be left as method stubs. I want this show its really clear for S3 implementor how to implement both history and visibility archival.

Reply:
Done.